### PR TITLE
1.0.0

### PR DESCRIPTION
**Pull request by denysdovhan**
Here we are working on **gulp-remark**.

### CODE REVIEW

**Comment by denysdovhan** [comment]
cc/ @wooorm @eush77 @iamstarkov

Any suggestions?

**Comment by wooorm** [comment]
Other than [mdastrc(5)](https://github.com/wooorm/mdast/blob/master/doc/mdastrc.5.md) and [mdastignore(5)](https://github.com/wooorm/mdast/blob/master/doc/mdastignore.5.md) support, nope. Looks good! :ok_hand:

**Comment by denysdovhan** [comment]
@wooorm is there any api how can i pass `.mdastrc` and `.mdastignore` configs into mdast?

**Comment by wooorm** [comment]
You’ll need to integrate with [mdast(1)](https://github.com/wooorm/mdast/tree/master/lib/cli/) for that, which isn’t very simple I guess.

Best start is the [file-set-pipeline](https://github.com/wooorm/mdast/blob/master/lib/cli/file-set-pipeline/index.js), which processes a collection of files. You’ll probably want to use its dependencies (most notably, [configure](https://github.com/wooorm/mdast/blob/master/lib/cli/file-set-pipeline/configure.js), [file-system](https://github.com/wooorm/mdast/blob/master/lib/cli/file-set-pipeline/file-system.js), and [log](https://github.com/wooorm/mdast/blob/master/lib/cli/file-set-pipeline/log.js)). There’s a lot going on in mdast(1) so I can imagine you’ll have to take some time to get familiar with the code, but it would be very powerful if gulp-mdast would integrate with the mdast(1), not just with mdast(3).

Before I forget, a lot of plug-ins expose messages, e.g., [mdast-lint](https://github.com/wooorm/mdast-lint). It would be very useful to have gulp-mdast expose those messages somewhere (either using the aforementioned `log` module or using [`vfile-reporter`](https://github.com/wooorm/vfile-reporter) directly).

P.S. [mdast(1) == CLI, mdast(3) == API](http://unix.stackexchange.com/questions/3586/what-do-the-numbers-in-a-man-page-mean).

**Comment by denysdovhan** [comment]
Is there possible way to require it from mdast?

For examle, how it was done in [`gulp-eslint`](https://github.com/adametry/gulp-eslint/blob/master/index.js#L5):

``` js
var CLIEngine = require('eslint').CLIEngine;
…
var linter = new CLIEngine(options);
```

Maybe could you create api for these cases (I mean grunt plugin, webpack loader e.g)? Because it seems to be a bit difficult.

**Comment by wooorm** [comment]
Of course it’s require-able: `require('mdast/lib/cli')`, for example, would work just fine!

As mentioned earlier, this is indeed a bit advanced. There are no external documents to guide you.
Don’t let that discourage you, though. There are a lot of docs inline. And, to be fair, there’s also a lot of complexity going on, which is the reason why it might seem complex :wink:

I’d appreciate it if you’d give it a go, but don’t fret to ask me if you run into things.
Also, I _may_ have some time tomorrow to try to whip something up! :smile:

**Comment by denysdovhan** [comment]
Oh, thank you. Have you got invite into this repo?

**Comment by wooorm** [comment]
Yup, thanks :+1:

**Comment by denysdovhan** [comment]
Ok. Will dive up into sources.

**Comment by denysdovhan** [comment]
There is a problem. I noticed it just now. `gulp-mdast` is [already on npm](https://www.npmjs.com/package/gulp-mdast).

What should we do?

**Comment by wooorm** [comment]
Seems @shinnn isn’t maintaining it (am I correct, Shinnosuke?) In which case, this project should be able to continue from version 0.0.4.

**Comment by wooorm** [comment]
Just saw `shinnn/gulp-mdast` was published today, just not on GH yet. Meaning, the same work happens  in two places, which is generally a bad idea :smile:

**Comment by shinnn** [comment]
@denysdovhan @wooorm 

It's OK to add @denysdovhan to the npm owners. I guess @denysdovhan can update gulp-mdast project more frequently because I'm already maintaining lots of other OSS projects.

File: README.md [hunk 1]
```diff
@@ -5,44 +5,64 @@
...
+
+#### options
+
+Type: `Object`  
+Default: `{}`
+
+See the [`mdast.process`][mdast-opts] options.
 
-#### input
+### gulpMdast.use([plugin][mdast-plugins][, options])
```
Comment (shinnn) [inline README.md]: `.use` method doesn't work as documented.

``` js
// doesn't work
gulpMdast.use(plugin);
```

``` js
// this works
gulpMdast().use(plugin);
```

Comment (denysdovhan) [inline README.md]: Sure. That's my mistake.

**Comment by wooorm** [comment]
Thanks, @shinnn. I hope that’s the best for gulp-mdast :smile: Any suggestions you came across when creating `shinnn/gulp-mdast` we should keep in mind?

P.S. Just checked npm; am I right that you haven’t added @denysdovhan yet? No haste, just making sure you haven’t forgotten :wink:

**Comment by shinnn** [comment]
> Just checked npm; am I right that you haven’t added @denysdovhan yet?

@wooorm, I just added @denysdovhan and you.

**Comment by wooorm** [comment]
:ok_hand:

**Comment by denysdovhan** [comment]
So will we continue in my repo or wherever e

**Comment by wooorm** [comment]
@denysdovhan Well, this repo is on GH. So I’d say here. Just ensure, when publishing, you start at `0.0.4`.

**Comment by denysdovhan** [comment]
I'd say do it here too. @shinnn, do you want to be added into this repo?

>  … you start at 0.0.4.

Maybe better to start from `0.1.0` or even `1.0.0`. We're changing API using `gulpMdast().use()` method, therefore, according to semver, we should bump major version. What do you think?

**Comment by iamstarkov** [comment]
semver doesnt dictate to do anything until 1.0

**Comment by iamstarkov** [comment]
I would say 0.1 is good enough to test API and find obvious bugs

**Comment by wooorm** [comment]
@denysdovhan I completely agree regarding semver, I was responding to the fact that there are two separate gulp-mdast’s, but that won’t pose a problem if all versions used by `denysdovhan/gulp-mdast` are higher than `shinnn/gulp-mdast`. :wink:

**Comment by denysdovhan** [comment]
ok. I'll continue from `0.1`

**Comment by shinnn** [comment]
https://github.com/denysdovhan/gulp-mdast/pull/1#issuecomment-162631153

> do you want to be added into this repo?

Of course! Thanks :)

**Comment by denysdovhan** [comment]
@shinnn You're invited. Sorry for late. I'm a little busy these days

**Comment by denysdovhan** [comment]
I'm not sure how should API look like.

`gulpRemark([options])` and attaching plugins through `gulpRemark.use(plugin[, options])` — is it good API?

**Comment by denysdovhan** [comment]
@wooorm can I somehow call [.use(plugin)](https://github.com/wooorm/remark/blob/5a96751edabf2c0654c33611b87eb047a779ca6d/lib/cli/file-set.js#L135-L172) and pass this `FileSet` into [.transform(context)](https://github.com/wooorm/remark/blob/master/lib/cli/file-set-pipeline/transform.js#L31)? Or pushing plugins into `context.plugins` it's the only way?

**Comment by wooorm** [comment]
@denysdovhan Nope, the `use` function on `file-set` is only applicable for file-sets and [`completer`](https://github.com/wooorm/remark/blob/master/doc/remarkplugin.3.md#function-completerfileset-next) functions, not for normal plug-ins.

Currently, it’s not possible to inject functions to be used as plug-ins on the CLI, only [strings on the `context` object](https://github.com/wooorm/remark/blob/5a96751edabf2c0654c33611b87eb047a779ca6d/lib/cli/configuration.js#L236) are injected into the from `.remarkrc` files detected [`plugins` object](https://github.com/wooorm/remark/blob/5a96751edabf2c0654c33611b87eb047a779ca6d/lib/cli/file-pipeline/configure.js#L177).

If, in the future, we’re planning on allowing the injection of plug-ins, it should be somewhere close to [here](https://github.com/wooorm/remark/blob/5a96751edabf2c0654c33611b87eb047a779ca6d/lib/cli/file-pipeline/configure.js#L177) I think.

**Comment by denysdovhan** [comment]
@wooorm as I understood, we can pass plugins only like this:

``` js
gulp.task("default", () =>
  gulp.src('*.md')
    .pipe(gulpRemark({
      plugins: {
        "lint": {
          "code-block-style": false
        }
      }
    }))
);
```

but it looks terrible. I wanna make something like this:

``` js
gulp.task("default", () =>
  gulp.src('*.md')
    .pipe(gulpRemark().use(lint, { "code-block-style": false }))
);
```

It's better, isn't it? Any idea how can I implement this? 

---

> Currently, it’s not possible to inject functions to be used as plug-ins on the CLI

If there isn't possible way to do this, maybe should we make it possible before?

**Comment by wooorm** [comment]
@denysdovhan I’ve added support for injecting plugins into remark(3) from remark(1). Read more about it: https://github.com/wooorm/remark/commit/4e65d7086041240047991c3f459e21811983978a.

Released in remark@3.1.0.

**Comment by denysdovhan** [comment]
@wooorm You have an mistake in https://github.com/wooorm/remark/commit/4e65d7086041240047991c3f459e21811983978a commit message:

``` js
var CLI = require('remark/lib/cli/cli');
var engine = require('remark/lib/cli');
var html = require('remark-html');

var cli = new CLI({
    'files': 'readme.md'
}).use(html);

run(cli, function (err, success) {
    console.log('done!', err, 'successfully?', success);
});
```

You required `engine`, but use `run()` which isn't required. By the way, [`remark/lib/cli`](https://github.com/wooorm/remark/blob/4e65d7086041240047991c3f459e21811983978a/lib/cli/index.js#L131) doesn't export `run()` function. Or maybe I'm wrong?

This makes me confused, so I still don't understand how may I use it.

**Comment by wooorm** [comment]
I changed the name of `engine` halfway to `run`, which in this case made more sense. Just replace both with something else, e.g., `whoops`, and stuff will work ;)

**Comment by denysdovhan** [comment]
Okay, got it) 

If understood correctly, now I can not to use [`configure()` and `transform()`](https://github.com/denysdovhan/gulp-remark/blob/feat/base/index.js#L28-L52). That means I can just pass `context` into `run()`, right?

**Comment by wooorm** [comment]
Close. The example in the commit wasn’t meant as being a replacement for this project though. What that commit does, is add support for inject plugins (as functions) into the CLI.

The following code is now supported:

``` js
var cli = new CLI({
    'files': 'readme.md'
}).use(html);
```

This means you could a bit of the code in gulp-remark, [specifically, L28-L36](https://github.com/denysdovhan/gulp-remark/blob/d9648439f38efe5e9f1d1c96b7aa58c61df3ef8f/index.js#L28-L36):

``` js
    context = {
      detectRC: true,
      cwd: process.cwd(),
      stdout: process.stdout,
      stderr: process.stderr,
      files: [
        vFile
      ]
    };
```

...with something like the first example.

E.g., the following:

``` js
var cli = new CLI({
  detectRC: true,
});

cli.files = [vFile];
cli.use(plugin);
```

Good luck!

**Comment by denysdovhan** [comment]
Thank you, I think I've got it.

**Comment by denysdovhan** [comment]
Hohoho!

It's almost done! Works with `.remarkignore` and `.remarkrc`, prints linter's reports and  processes files!

image

Now, I'm gonna write some tests and new readme. Does anybody have any suggestions?

**UPD:** Woops. Exactly it still doesn't read ignore files :(

**Comment by denysdovhan** [comment]
@wooorm one more question. If I understood correctly, `remark/lib/cli/ignore.js` just [read folder](https://github.com/wooorm/remark/blob/d4f0c67019c4b28164d41d7b0c07806968d9df0f/lib/cli/ignore.js#L73) and ignore files using `.remarkignore`.  Am I right?

If I am, please, can you add checking by `VFile.filePath()`?

I ask you about it, because `cli.traverser` actually contains `Finder { ignore: [ 'README.md', 'node_modules/' ], … }`, but it's doesn't ignore anything:

image

Possibly it happens 'cause gulp load every file that match with pattern defined in `gulp.src()` (currently `*.md`), but remark does not ignore them.

**Comment by wooorm** [comment]
Which code produces the output shown in the image? The current last commit, or are there specific changes?

**Comment by denysdovhan** [comment]
I've just added:

``` js
console.log(cli.traverser);
```

before [L36](https://github.com/denysdovhan/gulp-remark/blob/a1c0d8354aaaaf250be09c17bea63bbe0b6dffed/index.js#L36) and change `.remarkignore` like this:

``` diff
- example.md
+ README.md
```

**Comment by denysdovhan** [comment]
@wooorm oh, and I've change [context](https://github.com/denysdovhan/gulp-remark/blob/a1c0d8354aaaaf250be09c17bea63bbe0b6dffed/index.js#L13-L18):

``` js
  const cli = new CLI({
    detectRC: true,
    detectIgnore: true,
    cwd:      process.cwd(),
    stdout:   process.stdout,
    stderr:   process.stderr
  });
```

**Comment by wooorm** [comment]
Unfortunately, it’s not possible to ignore inside the CLI, as the ignore is only used when searching for files by globs. In Gulp, you’re not searching for files, thus the ignore is never reached. There are also some other good reasons why this shouldn’t change, which I won’t dig into.

As a solution, however, you are able to use `cli.traverser.ignore.shouldIgnore(filePath)` inside gulp-remark to check whether the `filePath` of a `file` is inside the ignored patterns, since [remark@3.2.2](https://github.com/wooorm/remark/releases/tag/3.2.2). Hope that helps!

**Comment by denysdovhan** [comment]
Ok, I'm trying to make it works. I wanna clarify: does `cli.traverser.ignore.shouldIgnore(filePath)` compare with basename or with absolute name?

**Comment by denysdovhan** [comment]
[I've sent pull request](https://github.com/wooorm/vfile/pull/3) to `vFile` with added `vFile#fileBasename` method. @wooorm please take a look.

**Comment by denysdovhan** [comment]
OK, I've fixed tests and everything works.

But we have another problem. How can I pass into cli [remark options](https://github.com/wooorm/remark#remarkprocessvalue-options-done)? I've already seen [this](https://github.com/wooorm/remark/blob/master/lib/cli/cli.js#L304-L320), but definitely that's not what I'm looking for.

**Comment by wooorm** [comment]
Currently, that’s possible through [`settings`](https://github.com/wooorm/remark/blob/master/lib/cli/cli.js#L289), which expects a string.

To be honest, maybe we should first push 1.0.0 now, and add this later.
I don’t think it’s really blocking, because it’s possible to specify settings through `.remarkrc` files now.

Do you agree?

**Comment by denysdovhan** [comment]
Okay, so I'm gonna write some tests and new `README.md`.

Do you wanna release `1.0.0` or `0.1.0`?

**Comment by wooorm** [comment]
1.0.0, as it’s also higher than [the current](https://www.npmjs.com/package/gulp-remark).

**Comment by denysdovhan** [comment]
Oh, right. Okay, I hope I'll finish today.

**Comment by shinnn** [comment]
@andrepolischuk @denysdovhan @wooorm Can you give me write access to `gulp-remark` package?

Also, what is [this repo](https://github.com/andrepolischuk/gulp-remark)? Current [npm page](https://www.npmjs.com/package/gulp-remark) links to it.

**Comment by denysdovhan** [comment]
@shinnn it's a @andrepolischuk ’s repo. I've already sent you invite to this repo. If you're talking about npm package, you was primary owner.

UPD: Added you to npm owners.

**Comment by shinnn** [comment]
> Added you to npm owners.

Confirmed. Thanks :)

```
$ npm owner ls gulp-remark

andrepolischuk <andre.polischuk@gmail.com>
denysdovhan <email@denysdovhan.com>
shinnn <snnskwtnb@gmail.com>
wooorm <tituswormer@gmail.com>
```

**Comment by wooorm** [comment]
Yes, haha, seems many people want a gulp plug-in, so let's give it to them :)

**Comment by denysdovhan** [comment]
Done. @wooorm and @shinnn, please make review.

We are ready to release :tada:

File: README.md [hunk 2]
```diff
@@ -1,56 +1,73 @@
-# gulp-mdast
+# gulp-remark
 
 [![NPM version][npm-image]][npm-url]
 [![Build Status][travis-image]][travis-url]
 [![Coveralls Status][coveralls-image]][coveralls-url]
 [![Dependency Status][depstat-image]][depstat-url]
 
-> Gulp plugin for Mdast
+> Gulp plugin for [Remark][remark] — markdown processor powered by plugins
```
Comment (wooorm) [inline README.md]: Could you lower-case the `Remark` > `remark`? Nitpicking, I know.

File: README.md [hunk 3]
```diff
@@ -1,56 +1,73 @@
...
+    .pipe(gulp.dest('dist'))
+);
 ```
 
 ## API
 
-### gulpMdast(input, [options])
+### gulpRemark()
+
+Gulp plug-in. It processes your files through [Remark][remark]. If you want to define specified options, use [`.remarkrc`][remarkrc]. The ignoring of files available using [`.remarkignore`][remarkignore] file.
```
Comment (wooorm) [inline README.md]: Same casing here, `Remark` > `remark`, and probably other places :wink:

File: README.md [hunk 4]
```diff
@@ -1,56 +1,73 @@
...
-
-gulpMdast('unicorns'); // unicorns
+import gulp from 'gulp';
+import remark from 'gulp-remark';
+import html from 'remark-html';
+import lint from 'remark-html';
+
+gulp.task('default', () =>
+  gulp.src('*.md')
+    .pipe(remark().use(html).use(lint))
```
Comment (wooorm) [inline README.md:24]: This looks great! If I remember correctly, there used to be a `gulpfile` in this repo which ran remark on the project itself. Maybe it’s a good idea to use gulp-remark in this project?

File: README.md [hunk 5]
```diff
@@ -1,56 +1,73 @@
...
 ## Usage
 
 ```js
-import gulpMdast from 'gulp-mdast';
-
-gulpMdast('unicorns'); // unicorns
+import gulp from 'gulp';
+import remark from 'gulp-remark';
+import html from 'remark-html';
+import lint from 'remark-html';
```
Comment (wooorm) [inline README.md]: `"remark-html"` > `"remark-lint"`

**Comment by wooorm** [comment]
Other than the comments, it looks great :+1:

File: README.md (hunk 2 above)
Comment (denysdovhan) [inline README.md]: OK, as you wish ;)

File: README.md (hunk 4 above)
Comment (denysdovhan) [inline README.md:24]: For what? I would be good as example, but personally I think npm scripts do this job well enough.

File: README.md (hunk 5 above)
Comment (denysdovhan) [inline README.md]: Woops

**Comment by denysdovhan** [comment]
Done!

**Comment by denysdovhan** [comment]
Well, does anyone have any remarks?) May I merge?

**Comment by wooorm** [comment]
I’m A-OK!

**Comment by denysdovhan** [comment]
Merged! Finally!

**Comment by denysdovhan** [comment]
Released as [`1.0.0`](https://www.npmjs.com/package/gulp-remark).

**Comment by denysdovhan** [comment]
@shinnn I'm gonna deprecate your `gulp-mdast` package, or maybe do you wanna do this?

**Comment by shinnn** [comment]
@denysdovhan I'll deprecate it.

**Comment by denysdovhan** [comment]
👍

**Pull request merged**